src/java.desktop/share/classes/com/sun/java/swing/SwingUtilities3.java

```diff
@@ -135,4 +143,72 @@ public static RepaintManager getDelegateRepaintManager(Component
         }
         return delegate;
     }
+
+    @FunctionalInterface
+    public interface PaintInterface {
+        void paintUnscaledBorder(Component c, Graphics g, int x, int y,
+                                 int w, int h, double scale, int strokeWidth);
+    }
+
+    public static void paintBorder(Component c, Graphics g, int x, int y,
+                            int w, int h, PaintInterface paintFunction) {
+
+        //STEP 1: RESET TRANSFORM
+        //save old values
+        AffineTransform at = null;
+        Stroke oldStk = null;
+        boolean resetTransform = false;
+        int stkWidth = 1;
+        double scaleFactor = 1;
+
+        if (g instanceof Graphics2D g2d) {
+            at = g2d.getTransform();
+            oldStk = g2d.getStroke();
+            scaleFactor = Math.min(at.getScaleX(), at.getScaleY());
+
+            /* Deactivate the HiDPI scaling transform,
+             * so we can do paint operations in the device
+             * pixel coordinate system instead of the logical coordinate system.
+             */
+            resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
+
+            if (resetTransform) {
+                g2d.setTransform(new AffineTransform());
+                stkWidth = c instanceof JInternalFrame ?
+                        clipRound(scaleFactor) : (int) Math.floor(scaleFactor);
+                g2d.setStroke(new BasicStroke((float) stkWidth));
```

Comment:
This is supposed to be a generic helper. If `strokeWidth` depends on the type of component, it's better to handle it in each component separately.

`EtchedBorder` uses the stoke, `LineBorder` doesn't use it; however, it uses the value.

So, my suggestion is to calculate the `strokeWidth` (can't we spell it out, it's easier to read this way, isn't it?) as `Math.floor(scaleFactor)`. The stroke itself isn't created, nor is set here.

The stroke can still be restored here as a convenience. What do others think?